### ISSUE TITLE
verification failures not counted

### ISSUE THREAD

**Issue by nniles** [fredericbonnet/picotest#1]
Calls to PICOTEST_ASSERT() increment the accumulated failure count, while calls to PICOTEST_VERIFY() do not.

I have augmented hooks.c with the failure logger from logger.c to make it obvious where this happens in the example output below.
```
running test suite mainSuite (4 subtests)...
  subtest testCase1 (1 out of 4)...
    running test case testCase1...
[VERIFY] /home/nate/projects/picotest/examples/mainSuite.inc(125) : FAILS
[ASSERT] /home/nate/projects/picotest/examples/mainSuite.inc(126) : FAILS
    test case testCase1 done (1 failures)
  => subtest testCase1 done (1 failed)...
  subtest testCase2 (2 out of 4)...
    running test case testCase2...
      fixture simpleFixture setup... done
[VERIFY] /home/nate/projects/picotest/examples/mainSuite.inc(134) : FAILS
      fixture simpleFixture teardown... done
    test case testCase2 done (0 failures)
  => subtest testCase2 done (0 failed)...
  subtest subSuite (3 out of 4)...
    running test suite subSuite (2 subtests)...
      subtest testCase4 (1 out of 2)...
        running test case testCase4...
          fixture simpleFixture setup... done
          fixture simpleFixture teardown... done
        test case testCase4 done (0 failures)
      => subtest testCase4 done (0 failed)...
      subtest testCase5 (2 out of 2)...
        running test case testCase5...
          fixture fixtureWithOptionalContext setup... done
          fixture fixtureWithOptionalContext teardown... done
        test case testCase5 done (0 failures)
      => subtest testCase5 done (0 failed)...
    test suite subSuite done (0 failures)
  => subtest subSuite done (0 failed)...
  subtest testCase3 (4 out of 4)...
    running test case testCase3...
      fixture fixtureWithContext setup... done
[VERIFY] /home/nate/projects/picotest/examples/mainSuite.inc(143) : context->var2 == NULL
[ASSERT] /home/nate/projects/picotest/examples/mainSuite.inc(144) : context->var3 == NULL
      fixture fixtureWithContext teardown... done
    test case testCase3 done (1 failures)
  => subtest testCase3 done (1 failed)...
test suite mainSuite done (2 failures)
```
The failure logger was called 5 times (2 asserts, 3 verifies), but the failure tally for the suite only indicates 2 failures, equal to the number of failed assertions.

Digging through the code, it appears that the failure accumulation mechanism is tied to the longjmp() path PICOTEST_ABORT(), so maybe this is a fundamental issue?

**Comment by nniles** [comment]
Patch attached.

[count-verify-failures-patch.txt](https://github.com/fredericbonnet/picotest/files/3826921/count-verify-failures-patch.txt)

**Comment by fredericbonnet** [comment]
Hi Nathaniel, first of all thanks for using my package, and thanks for your patch as well!

The way `PICOTEST_VERIFY()` soft assert works in the current version is as I originally intended. When I designed PicoTest, I thought that only `PICOTEST_ASSERT()` hard asserts should be reported as failures, because such tests don't fail in the strict sense of the term (they run to completion). However, as you point out, this breaks the semantics of the failure logger (there should be exactly one call per failure). So I wouldn't make the same choice today. This means we need a way for PicoTest to report the failed soft assert count at the end of the tests. There are two ways to do so:

1. Implicitly, like your patch does.
2. Explicitly, like many frameworks do (for example TestNG `SoftAssert.assertAll()` or AssertJ `SoftAssertions.assertAll()`). This would require e.g. calling `PICOTEST_ASSERT_ALL()` at the end of the test.

Option 2 is too verbose IMHO and your patch looks fine so I will merge it in the next version, with the appropriate version notice (potential but minor incompatibility).

Thanks again!

**Comment by nniles** [comment]
Sorry, I should have started by thanking you for sharing this great little unit testing framework.

I guess I have a slightly different take on verify vs. assert, in that I use assert only where running subsequent tests would break something in the test code, and/or render further results meaningless.  E.g., a search returns an empty set when some data was expected and is required to perform further testing.

Thanks, and thanks for taking my patch!

**Comment by fredericbonnet** [comment]
No worries! I've just released version 1.1.0 that fixes the issue. I couldn't merge your patch as is because it only worked when calling assertions directly from within test cases and not indirectly from called functions (such functions have no access to the implicit `fail` argument), but otherwise it works the same.
